blog/2024-03-08-structuring-a-large-app.md

```diff
@@ -0,0 +1,391 @@
+---
+author: Masen Furer
+date: 2024-03-08
+title: "Structuring a Large App"
+description: "So your Reflex app is getting large? Here's some advice on how to lay it out."
+image: /blog/structure-large-app.jpg
+---
+
+# First Steps
+
+Reflex makes it easy to get a simple web app running from a single python file:
+
+```python
+import reflex as rx
+
+
+@rx.page()
+def index():
+    return rx.center(
+        rx.heading("Hello Reflex! 👋")
+    )
+
+
+app = rx.App()
+```
+
+However, as the app grows, keeping everything in a single module can quickly
+become messy and unmanageable. Because a Reflex app is just a python project,
+it can be organized into modules and packages (directory with `__init__.py`)
+and use the python import machinery you're likely already familiar with.
+
+# What Makes a Reflex App?
+
+When executing `reflex init` or `reflex run`, the framework will import **two** modules automatically.
+
+## rxconfig.py
+
+Reflex first looks for an `rxconfig.py` in the current working directory and
+imports it.  At minimum, it **must define a module-level global named `config` as
+an instance of `rx.Config`**, of which the only required setting is `app_name`.
+
+```python
+import reflex as rx
+
+config = rx.Config(app_name="example_big_app")
+```
+
+## \{app_name}/\{app_name}.py
+
+After loading the config, Reflex imports the main app module based on the
+`app_name`, which **must define a module-level global named `app` as an instance
+of `rx.App`**.  In the example above, this would be
+`example_big_app/example_big_app.py`.
+
+The main app module is responsible for importing all other modules that make up the
+app and defining `app = rx.App()`. **All other modules containing pages, state,
+and models MUST be imported by the main app module or package** for Reflex to be able to
+"see" these parts of the app.
+
+# How to Split Up the App
+
+The following tree is a recommended structure for breaking a Reflex app up into logic pieces.
+
+```text
+example-big-app/
+├─ assets/
+├─ example_big_app/
+│  ├─ components/
+│  │  ├─ __init__.py
+│  │  ├─ auth.py
+│  │  ├─ footer.py
+│  │  ├─ menu.py
+│  │  ├─ navbar.py
+│  ├─ pages/
+│  │  ├─ __init__.py
+│  │  ├─ index.py
+│  │  ├─ login.py
+│  │  ├─ posts.py
+│  │  ├─ product.py
+│  │  ├─ profile.py
+│  │  ├─ schedule.py
+│  ├─ __init__.py
+│  ├─ example_big_app.py
+│  ├─ models.py
+│  ├─ state.py
+│  ├─ template.py
+├─ uploaded_files/
+├─ requirements.txt
+├─ rxconfig.py
+```
+
+## example_big_app/models.py
```

Comment:
I would start with the pages (then components, then models) maybe. Just to go from most used feature to least used feature.